modules/core/src/main/java/org/apache/ignite/cache/validation/IgnitePluggableSegmentationResolver.java

```diff
@@ -0,0 +1,275 @@
+        Boolean res = segResolverEnabledProp.get();
+
+        return res == null || !res;
+    }
+
+    /** @return return. */
+    private String formatTopologyNodes(Collection<ClusterNode> nodes) {
+        return nodes.stream().map(n -> n.id().toString()).collect(Collectors.joining(", "));
+    }
+
+    /** */
+    private class TopologyChangedEventListener implements DiscoveryEventListener, HighPriorityListener {
+        /** {@inheritDoc} */
+        @Override public void onEvent(DiscoveryEvent evt, DiscoCache discoCache) {
+            lastCheckedTopVer = evt.topologyVersion();
+
+            if (isDisabled())
+                return;
+
+            if (state == State.VALID && evt.type() == EVT_NODE_FAILED) {
+                List<? extends BaselineNode> baselineNodes = discoCache.baselineNodes();
+
+                if (baselineNodes != null && aliveBaselineNodes(baselineNodes) < baselineNodes.size() / 2 + 1) {
+                    state = State.INVALID;
+
+                    stateChangeExec.execute(() -> {
+                        try {
+                            ctx.cluster().get().state(ACTIVE_READ_ONLY);
+                        }
+                        catch (Throwable e) {
+                            U.error(
+                                log,
+                                "Failed to automatically switch state of the segmented cluster to the READ-ONLY mode" +
+                                    " [segmentedNodes=" + formatTopologyNodes(discoCache.allNodes()) + "]. Cache writes" +
+                                    " are already restricted for all configured caches, but this step is still required" +
+                                    " in order to be able to unlock cache writes in the future. Retry this operation" +
+                                    " manually, if possible.",
+                                e
+                            );
+                        }
+                    });
+
+                    U.warn(log, "Cluster segmentation was detected [segmentedNodes=" +
+                        formatTopologyNodes(discoCache.allNodes()) + ']');
+                }
+            }
+
+            if (ctx.state().isBaselineAutoAdjustEnabled())
+                U.warn(log, "Segmentation Resolver requires baseline to be configured. If no baseline is" +
+                    " set, any topology change is considered valid.");
+        }
+
+        /** {@inheritDoc} */
+        @Override public int order() {
+            return 0;
+        }
+
+        /** */
+        private int aliveBaselineNodes(Collection<? extends BaselineNode> baselineNodes) {
+            int res = 0;
+
+            for (BaselineNode node : baselineNodes) {
+                if (!(node instanceof DetachedClusterNode))
+                    ++res;
+            }
+
+            return res;
+        }
+    }
+
+    /** */
+    private class ClusterStateChangedEventListener implements CustomEventListener<ChangeGlobalStateFinishMessage> {
```

Comment:
From my point of view, it would be better to "inline" this class.

Reply:
But in this case we will crumble up the `start` method and won't  gain any real profit.